### PR TITLE
chore: amd64 and arm64 use PIE build mode by default

### PR DESCRIPTION
**Pull request by shenghuang147**
### Usage:
``````shell
make BUILD_MODE='-buildmode=(build mode)'
# or
BUILD_MODE='-buildmode=(build mode)' make
``````

### Added the feature of setting BUILD_MODE according to GOARCH in Makefile

I added some code in Makefile to set different BUILD_MODE values  according to different GOARCH values, so as to achieve cross-platform  compilation and position-independent executable files. This allows the  program to run on different operating systems and computing  architectures, and also improves the security of the program. This PR is to enhance the compatibility and stability of the project.

I specifically specified that only when GOARCH is amd64 or arm64,  -buildmode=pie will be used to set BUILD_MODE. This is because pie may  not be supported on other computing architectures. Therefore, I used an  ifeq statement to judge the value of GOARCH and set BUILD_MODE according to different situations.

I also added a feature that allows users to manually specify the  build mode using make BUILD_MODE=-buildmode=*, overriding the default  settings. This allows users to choose the appropriate build mode  according to their needs, such as , pie, c-shared, c-archive,  shared, plugin, etc. If the user does not manually specify the build  mode, then if GOARCH is amd64 or arm64, -buildmode=pie will be used.

### Checklist

- [x] The Pull Request has been fully tested
- [ ] There's an entry in the CHANGELOGS
- [ ] There is a user-facing docs PR against https://github.com/daeuniverse/dae-wing

### Full Changelogs

- [Implement ...]

### Issue Reference

Fix #_[issue number]_

### Test Result

I used different GOARCH and BUILD_MODE values to compile the project  and run tests on amd64 and arm64 platforms. The test results show that  my modification can correctly set BUILD_MODE and does not affect the  normal operation of other parameters and functions. 

**test code:**

``````shell
#!/bin/bash

make
echo ""
echo ""

for arch in 386 amd64 arm arm64 mips mips64 mips64le mipsle; do
        echo "===================$arch=========================="
        make GOARCH=$arch
        echo ""
        echo "============$arch Test Set BUILD_MODE============="
        make VERSION=$arch BUILD_MODE=-buildmode=exe
        echo ""


done
``````

**result:**

``````shell
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=pie  .


===================386==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." '   .

============386 Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=386 -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================amd64==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=pie  .

============amd64 Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=amd64 -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================arm==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." '   .

============arm Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=arm -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================arm64==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=pie  .

============arm64 Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=arm64 -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================mips==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." '   .

============mips Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=mips -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================mips64==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." '   .

============mips64 Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=mips64 -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================mips64le==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." '   .

============mips64le Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=mips64le -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .

===================mipsle==========================
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=0.0.0.unknown -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." '   .

============mipsle Test Set BUILD_MODE=============
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
WARN[0000] dangerous converting: may exceeds graphQL int32 range  name=SoMarkFromDae type=uint32
go build -o ./dae-wing -trimpath -ldflags '-s -w -X github.com/daeuniverse/dae-wing/db.AppVersion=mipsle -X github.com/daeuniverse/dae-wing/db.AppName=dae-wing -X "github.com/daeuniverse/dae-wing/db.AppDescription=dae-wing is a integration solution of dae, API and UI." ' -buildmode=exe  .
``````

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -70,7 +76,7 @@ bundle: deps
 				rm "{}"; \
 			fi' ';' ; \
 	fi && \
-	go build -tags=embedallowed -o $(OUTPUT) -trimpath -buildmode=pie -ldflags $(LDFLAGS) .
+	go build -tags=embedallowed -o $(OUTPUT) -trimpath -ldflags $(LDFLAGS) $(BUILD_MODE) .
```
Comment (kunish) [inline Makefile:79]: BUILD_MODE is not that explicit here, you can

either

```shell
go build [...] -buildmode=$(BUILD_MODE)
```

or

```shell
go build [...] $(EXTRA_BUILD_ARGS)
```

Comment (mzz2017) [reply Makefile:79]: -buildmode=$(BUILD_MODE) may throw error if BUILD_MODE is empty.

Comment (mzz2017) [reply Makefile:79]: I didn't understand your point. Can you explain it in more detail? @kunish

Comment (kunish) [reply Makefile:79]: > -buildmode=$(BUILD_MODE) may throw error if BUILD_MODE is empty.

If that's the case, we can set the default value of BUILD_MODE as `default`. See `go help buildmode`

image

Comment (kunish) [reply Makefile:79]: > I didn't understand your point. Can you explain it in more detail? @kunish

The point is if i'm a user of this makefile, the design of this `BUILD_MODE` variable does not feel very appealing and obvious to me.

Comment (kunish) [reply Makefile:79]: To supplement my suggestions above

```shell
make BUILD_MODE=pie
```

```shell
make EXTRA_BUILD_ARGS="-buildmode=pie"
```

Comment (mzz2017) [reply Makefile:79]: This is not designed for users. It is an internal variable in most cases unless the user wants to overwrite it.

Comment (kunish) [reply Makefile]: > This is not designed for users. It is an internal variable in most cases unless the user wants to overwrite it.

I think it's not a good design either from a developer perspective.

Comment (mzz2017) [reply Makefile]: See https://github.com/daeuniverse/dae-wing/pull/102/commits/616f3e560a0cdd734d7ad64578778e35829004dd

**Review by mzz2017 — APPROVED** [review]

**Review by kunish — APPROVED** [review]
LGTM

**Comment by kunish** [comment]
Since this repo does not have a PR build check yet, we can merge it for now and test it in daed

**Pull request merged**